Add missing type hints to src/hiero_sdk_python/file/file_append_transaction.py

🆕🐥 Good first Issue

This issue is reserved for people who have never or only recently contributed to [Hiero](https://hiero.org) or any open source project in general. We know that creating a pull request (PR) is a major barrier for new contributors.
The goal of this issue and all other issues labeled by [**'good first issue'**](https://github.com/issues?q=is%3Aopen+is%3Aissue+org%3Ahiero-ledger+archived%3Afalse+label%3A%22good+first+issue%22+) is to help you make your first contribution to Hiero.

## 👾 Description of the issue

file_append_transaction.py appends data to an existing file on the network, and has some type hinting but the type hinting can be improved.

⚠️ Missing / Incomplete Type Hints
Running mypy on strict mode throws up for file_append_transaction:
mypy --strict src/hiero_sdk_python/file/file_append_transaction.py

(ignore the type conflict issues in this PR or those in other files)

src/hiero_sdk_python/file/file_append_transaction.py:58: error: Call to untyped function "to_tinybars" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:64: error: Missing type parameters for generic type "list"  [type-arg]
src/hiero_sdk_python/file/file_append_transaction.py:161: error: Function is missing a return type annotation  [no-untyped-def]
src/hiero_sdk_python/file/file_append_transaction.py:187: error: Function is missing a return type annotation  [no-untyped-def]
src/hiero_sdk_python/file/file_append_transaction.py:194: error: Call to untyped function "_build_proto_body" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:206: error: Call to untyped function "_build_proto_body" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:229: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
src/hiero_sdk_python/file/file_append_transaction.py:245: error: Function is missing a return type annotation  [no-untyped-def]
src/hiero_sdk_python/file/file_append_transaction.py:245: note: Use "-> None" if function does not return a value
src/hiero_sdk_python/file/file_append_transaction.py:259: error: Function is missing a type annotation  [no-untyped-def]
src/hiero_sdk_python/file/file_append_transaction.py:280: error: "None" has no attribute "valid_start"  [attr-defined]
src/hiero_sdk_python/file/file_append_transaction.py:294: error: "None" has no attribute "account_id"  [attr-defined]
src/hiero_sdk_python/file/file_append_transaction.py:304: error: Call to untyped function "build_transaction_body" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:313: error: Function is missing a type annotation  [no-untyped-def]
src/hiero_sdk_python/file/file_append_transaction.py:325: error: Call to untyped function "_validate_chunking" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:329: error: Call to untyped function "execute" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:345: error: Call to untyped function "freeze_with" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:350: error: Call to untyped function "sign" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:354: error: Call to untyped function "execute" in typed context  [no-untyped-call]
src/hiero_sdk_python/file/file_append_transaction.py:360: error: Function is missing a type annotation  [no-untyped-def]
src/hiero_sdk_python/file/file_append_transaction.py:377: error: Call to untyped function "sign" in typed context  [no-untyped-call]

Therefore, we should improve the type hinting.
Not all of these are feasible to solve in this PR, just a little general improvement is enough!

## 💡 Solution
We should type hint the functions, such as the argument and return types.

For example:
def _build_proto_body(self) -> file_append_pb2.FileAppendTransactionBody:

from hiero_sdk_python.hapi.services import transaction_body_pb2

def build_transaction_body(self) -> transaction_body_pb2.TransactionBody:

def _from_proto(self, proto: file_append_pb2.FileAppendTransactionBody) -> 'FileAppendTransaction':

from hiero_sdk_python.client import Client

def freeze_with(self, client: Client) -> 'FileAppendTransaction':
def execute(self, client: Client) -> TransactionReceipt:

from hiero_sdk_python.keys import PrivateKey

def sign(self, private_key: PrivateKey) -> 'FileAppendTransaction':

def _get_method(self, channel: _Channel) -> _Method[file_append_pb2.FileAppendTransactionBody, None]:

This should reduce down some of the mypy issues.

You can follow the examples of other type hinting in the SDK


## Acceptance Criteria:
- [ ] Most functions are type hinted
- [ ] Mypy strict conflicts reduce significantly
- [ ] All tests pass

### Implementation
- [ ] Add type hinting to arguments and return types in functions

### Additional requirements to Pass
- [ ] Short but descriptive changelog entry under 'UNRELEASED'
- [ ] Verified commits achieved (see CONTRIBUTING.md and signing.md)

### Next steps
  - [ ] Comment below that you’d like to work on this issue.
  - [ ] Wait to be assigned by a maintainer.
  - [ ] Read [CONTRIBUTING.md](https://github.com/hiero-ledger/hiero-sdk-python/blob/main/CONTRIBUTING.md)
 to start forking and developing.


If you have never contributed to an open source project at GitHub, the following step-by-step guide will introduce you to the workflow.
More information and concrete samples for shell commands for each step can be found in our [CONTRIBUTING.md](https://github.com/hiero-ledger/.github/blob/main/CONTRIBUTING.md)
 file.
A more detailed general documentation of the GitHub PR workflow can be found [here](https://github.com/firstcontributions/first-contributions/blob/master/README.md)
.

For Python-specific contributing guidelines, see:

[CONTRIBUTING.md](https://github.com/hiero-ledger/hiero-sdk-python/blob/main/CONTRIBUTING.md)

[README.md](https://github.com/hiero-ledger/hiero-sdk-python/blob/main/README.md)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,8 +6,8 @@ This changelog is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.
 
 ## [Unreleased]
 
-
 ### Added
+
 - Add Google-style docstrings to `AccountInfo` class and its methods in `account_info.py`.
 
 - add AccountRecordsQuery class
@@ -17,13 +17,14 @@ This changelog is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.
 ### Fixed
 
 - Added explicit read permissions to examples.yml (#623)
+- Improved type hinting in `file_append_transaction.py` to resolve 'mypy --strict` errors. ([#495](https://github.com/hiero-ledger/hiero-sdk-python/issues/495))
 
 ### Breaking Changes
-
 
 ## [0.1.7] - 2025-10-28
 
 ### Added
+
 - Expanded `README.md` with a new "Follow Us" section detailing how to watch, star, and fork the repository (#472).
 - Refactored `examples/topic_create.py` into modular functions for better readability and reuse.
 - Add Rebasing and Signing section to signing.md with instructions for maintaining commit verification during rebase operations (#556)
@@ -63,7 +64,7 @@ This changelog is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.
 - Converted monolithic function in `token_create_nft_infinite.py` to multiple modular functions for better structure and ease.
 - docs: Use relative paths for internal GitHub links (#560).
 - Update pyproject.toml maintainers list.
-– docs: Updated README.md/CHANGELOG.md and added blog.md, bud.md and setup.md (#474)
+  – docs: Updated README.md/CHANGELOG.md and added blog.md, bud.md and setup.md (#474)
 - renamed docs/sdk_developers/changelog.md to docs/sdk_developers/changelog_entry.md for clarity.
 - Refactor `query_balance.py` into modular, reusable functions with `setup_client()`, `create_account()`, `get_balance()`, `transfer_hbars()`, and `main()` for improved readability, maintainability, and error handling.
 - Unified balance and transfer logging format — both now consistently display values in hbars for clarity.
@@ -75,6 +76,7 @@ This changelog is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.
 - Type hinting for `Topic` related transactions.
 
 ### Removed
+
 - Remove deprecated camelCase alias support and `_DeprecatedAliasesMixin`; SDK now only exposes snake_case attributes for `NftId`, `TokenInfo`, and `TransactionReceipt`. (Issue #428)
 
 ## [0.1.6] - 2025-10-21
@@ -131,6 +133,7 @@ This changelog is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.
 ## [0.1.5] - 2025-09-25
 
 ### Added
+
 - ScheduleSignTransaction class
 - NodeUpdateTransaction class
 - NodeDeleteTransaction class

--- a/src/hiero_sdk_python/file/file_append_transaction.py
+++ b/src/hiero_sdk_python/file/file_append_transaction.py
@@ -1,3 +1,5 @@
+from __future__ import annotations
+
 """
 Represents a file append transaction on the network.
 
@@ -12,18 +14,25 @@ to build and execute a file append transaction.
 """
 
 import math
-from typing import Optional
+from typing import TYPE_CHECKING, Any, List, Optional
 from hiero_sdk_python.file.file_id import FileId
 from hiero_sdk_python.hbar import Hbar
 from hiero_sdk_python.transaction.transaction import Transaction
-from hiero_sdk_python.channels import _Channel
-from hiero_sdk_python.executable import _Method
 from hiero_sdk_python.transaction.transaction_id import TransactionId
 from hiero_sdk_python.hapi.services import file_append_pb2, timestamp_pb2
 from hiero_sdk_python.hapi.services.schedulable_transaction_body_pb2 import (
     SchedulableTransactionBody,
 )
 
+# Use TYPE_CHECKING to avoid circular import errors
+if TYPE_CHECKING:
+    from hiero_sdk_python.client import Client
+    from hiero_sdk_python.keys import PrivateKey
+    from hiero_sdk_python.channels import _Channel
+    from hiero_sdk_python.executable import _Method
+    from hiero_sdk_python.transaction_receipt import TransactionReceipt
+    
+    
 # pylint: disable=too-many-instance-attributes
 class FileAppendTransaction(Transaction):
     """
@@ -60,8 +69,8 @@ class FileAppendTransaction(Transaction):
         # Internal tracking for chunking
         self._current_chunk_index: int = 0
         self._total_chunks: int = self._calculate_total_chunks()
-        self._transaction_ids: list[TransactionId] = []
-        self._signing_keys: list = []  # Store all signing keys for multi-chunk transactions
+        self._transaction_ids: List[TransactionId] = []
+        self._signing_keys: List["PrivateKey"] = []  # Use string annotation to avoid import issues
 
     def _encode_contents(self, contents: Optional[str | bytes]) -> Optional[bytes]:
         """
@@ -99,7 +108,7 @@ class FileAppendTransaction(Transaction):
         """
         return self._calculate_total_chunks()
 
-    def set_file_id(self, file_id: FileId) -> 'FileAppendTransaction':
+    def set_file_id(self, file_id: FileId) -> FileAppendTransaction:
         """
         Sets the file ID for this file append transaction.
 
@@ -113,7 +122,7 @@ class FileAppendTransaction(Transaction):
         self.file_id = file_id
         return self
 
-    def set_contents(self, contents: Optional[str | bytes]) -> 'FileAppendTransaction':
+    def set_contents(self, contents: Optional[str | bytes]) -> FileAppendTransaction:
         """
         Sets the contents for this file append transaction.
 
@@ -129,7 +138,7 @@ class FileAppendTransaction(Transaction):
         self._total_chunks = self._calculate_total_chunks()
         return self
 
-    def set_max_chunks(self, max_chunks: int) -> 'FileAppendTransaction':
+    def set_max_chunks(self, max_chunks: int) -> FileAppendTransaction:
         """
         Sets the maximum number of chunks allowed for this transaction.
 
@@ -143,7 +152,7 @@ class FileAppendTransaction(Transaction):
         self.max_chunks = max_chunks
         return self
 
-    def set_chunk_size(self, chunk_size: int) -> 'FileAppendTransaction':
+    def set_chunk_size(self, chunk_size: int) -> FileAppendTransaction:
         """
         Sets the chunk size for this transaction.
 
@@ -158,7 +167,7 @@ class FileAppendTransaction(Transaction):
         self._total_chunks = self._calculate_total_chunks()
         return self
 
-    def _build_proto_body(self):
+    def _build_proto_body(self) -> file_append_pb2.FileAppendTransactionBody:
         """
         Returns the protobuf body for the file append transaction.
 
@@ -184,7 +193,7 @@ class FileAppendTransaction(Transaction):
             contents=chunk_contents
         )
 
-    def build_transaction_body(self):
+    def build_transaction_body(self) -> Any:
         """
         Builds the transaction body for this file append transaction.
 
@@ -208,7 +217,7 @@ class FileAppendTransaction(Transaction):
         schedulable_body.fileAppend.CopyFrom(file_append_body)
         return schedulable_body
 
-    def _get_method(self, channel: _Channel) -> _Method:
+    def _get_method(self, channel: "_Channel") -> "_Method":
         """
         Gets the method to execute the file append transaction.
 
@@ -221,12 +230,13 @@ class FileAppendTransaction(Transaction):
         Returns:
             _Method: An object containing the transaction function to append to a file.
         """
+        from hiero_sdk_python.executable import _Method
         return _Method(
             transaction_func=channel.file.appendContent,
             query_func=None
         )
 
-    def _from_proto(self, proto) -> 'FileAppendTransaction':
+    def _from_proto(self, proto: file_append_pb2.FileAppendTransactionBody) -> FileAppendTransaction:
         """
         Initializes a new FileAppendTransaction instance from a protobuf object.
 
@@ -242,7 +252,7 @@ class FileAppendTransaction(Transaction):
         self._total_chunks = self._calculate_total_chunks()
         return self
 
-    def _validate_chunking(self):
+    def _validate_chunking(self) -> None:
         """
         Validates that the transaction doesn't exceed the maximum number of chunks.
         
@@ -256,7 +266,7 @@ class FileAppendTransaction(Transaction):
             )
 
 
-    def freeze_with(self, client):
+    def freeze_with(self, client: "Client") -> FileAppendTransaction:
         """
         Freezes the transaction by building the transaction body and setting necessary IDs.
         
@@ -272,6 +282,7 @@ class FileAppendTransaction(Transaction):
         if self._transaction_body_bytes:
             return self
 
+        
         if self.transaction_id is None:
             self.transaction_id = client.generate_transaction_id()
 
@@ -310,7 +321,7 @@ class FileAppendTransaction(Transaction):
         return self
 
 
-    def execute(self, client):
+    def execute(self, client: "Client") -> Any:
         """
         Executes the file append transaction.
         
@@ -349,7 +360,6 @@ class FileAppendTransaction(Transaction):
                 # Call parent sign directly to avoid modifying _signing_keys
                 super().sign(signing_key)
 
-
             # Execute the chunk
             response = super().execute(client)
             responses.append(response)
@@ -357,7 +367,7 @@ class FileAppendTransaction(Transaction):
             # Return the first response (as per JavaScript implementation)
             return responses[0] if responses else None
 
-    def sign(self, private_key):
+    def sign(self, private_key: "PrivateKey") -> FileAppendTransaction:
         """
         Signs the transaction using the provided private key.
             
@@ -374,4 +384,5 @@ class FileAppendTransaction(Transaction):
             self._signing_keys.append(private_key)
 
         # Call the parent sign method for the current transaction
-        return super().sign(private_key)
+        super().sign(private_key)
+        return self

